### ISSUE TITLE
Shorten startup time for multiple workers

### ISSUE THREAD

**Issue by joe42** [benoitc/gunicorn#2693]
Hi,

When running our application with gunicorn with multiple workers I notice a considerable delay during startup.
All workers start in parallel and each worker takes an average of 5 sec to start.
What happens now with 8 workers is basically that the first worker completes its startup after roughly ~5*8 seconds instead of after 5 seconds.
Is there an easy way to make the worker startup sequential, like add a custom delay so that one of the workers can already start processing requests until the other workers are started? Or do I need to add some server hooks?

gunicorn 20.1.0
worker_class = 'gevent'
gevent 21.8.0
Python 3.9

**Comment by joe42** [comment]
I am setting the variable `preload_app = True` in the configuration file now, which significantly reduces startup time of multiple workers. 
However, this leads to an issue with gevent because the monkey-patching is done after the application code is loaded which is too late. That basically means all HTTPs requests inside the application fail, which is really bad.
So what you need to do is to manually call 

    from gevent import monkey
    monkey.patch_all()  # Monkey patch as early as possible.

As early as possible during your application startup.

I still think starting the workers one by one would be a really nice feature if that is not possible at the moment.
However, I am happy with using the application reload. 

It would be super nice if you could add that caveat in the configuration documentation for the preload parameter to reduce the amount of headaches in the world :)
Let me know if you want me to create a PR for that. I would be happy to help with that.

**Comment by benoitc** [comment]
> I am setting the variable `preload_app = True` in the configuration file now, which significantly reduces startup time of multiple workers. (See also: #1566) However, this leads to an issue with gevent because the monkey-patching is done after the application code is loaded which is too late. That basically means all HTTPs requests inside the application fail, which is really bad. So what you need to do is to manually call
> 
> ```
> from gevent import monkey
> monkey.patch_all()
> ```
> 
> Before your application startup. I put it into the gunicorn config for now, so it looks something like this:
> 
> ```
> from gevent import monkey
> monkey.patch_all()
> 
> worker_class = 'gevent'
> preload_app = True
> ```
> 
never do that. You will unblock the arbiter loop which will result in bad supervision. Is your code using global variable that need to be used before your application is loaded?

**Comment by jamadden** [comment]
> never do that. You will unblock the arbiter loop which will result in bad supervision. 

In our experience running a large production application and performing the gevent-monkey patch before even importing gunicorn, we have noticed no problems over the course of many years. cf https://github.com/benoitc/gunicorn/issues/1566

**Comment by Auric-Manteo** [comment]
@benoitc I haven't experienced any issue with the workaround so far. I would be interested to know more about the issue you describe and discuss a better way forward like an alternative way to doing the application pre-load or starting the workers one by one.

**Comment by benoitc** [comment]
no activity since awhile. closing feel free to create a new ticket if needed.
